exonum/src/runtime/mod.rs

```diff
@@ -68,6 +68,15 @@
 //!
 //! 6. After execution the transaction [execution status] is written into the blockchain.
 //!
+//! # Service interfaces
+//!
+//! In addition to its own methods, a service can implement methods of additional interfaces.
+//! In your own runtime implementation, you must provide support of the following interfaces:
+//!
+//! ## Initialize
```

Comment:
I remember from the conversation that the runtime was responsible to control the caller of some interfaces, requiring some special handling in the implementation, which negates any benefits from fewer number of methods. Is it still the case?

Reply:
Each service instance should validate caller in each method regardless of interface. In this case there are no special cases for the `Initialize` interface.

Reply:
I don't see anywhere that any validations are required (it didn't use to be the case!)

Reply:
For example you should verify that transaction author is authorized to transfer funds. In other words, anyway you have to write code like following.

```rust
let (author, tx_hash) = context.caller().as_transaction().ok_or(Error::WrongCaller)?;
if !is_authorized(context.fork(), &author) {
    return Err(Error::UnauthorizedOperation);
}
```

And even more safe way is write code like this:

```
let (author, fork) = context.verify_caller(|caller, snapshot| 
    caller.as_transaction().and_then(|(author, _)|
         if is_authorized(snapshot, author) { Some(author) }
         else { None }
    )
).ok_or(Error::SomeErrorType);
... logic
```

So i think you can write the same code in Java.

Reply:
> you should verify that transaction author is authorized to transfer funds

That is a natural responsibility of the application code to authorize the operation. We discuss the built-in "interface", and I don't see why either the runtime or the appication code has to perform any authorization of an operation that is supposed to be performed by the framework. It does not make sense to authorize the operations performed by the framework — why can't the framework itself ensure that this operation is performed at the right time and by the authorized entity (= the framework!)?

Reply:
Why would framework suppose that some caller is authorized (or not) for some kind of runtime or user application?

If framework will perform `some` checks, it will result in checking the same thing twice (some checks by the framework, and some by end-side code) and create a field to shoot yourself in a leg (by expecting the framework to perform all checks for you).

IMHO letting the end-side code check everything is OK.

Reply:
Because it is not an arbitrary operation, it is an operation that is suppossed to be performed by the **framework itself**. Isn't it ridiculous to have to check that your framework works as expected (or to enable misuse of the framework operations by implementing them in such a way)? Why don't we check then that 'deploy-artifact' is invoked by Exonum, and not some wonderful forces?

Reply:
I think that discussing in the abstract field won't lead to anything useful.

What's your concrete proposal?

Reply:
If you look at the start, there is nothing abstract.
1. Use a normal method that does not require fancy documentation and implementation instead of half-baked interfaces, or
1. Make the framework responsible for controlling the caller.

In the very least, if it can't be fixed, then document what the implementation shall do with this interface!

Reply:
1. That's not a proposal, just like "Do things the correct way and not how you're doing it now".
2. Ok, framework sees that `Caller::Transaction` attempts to call an user-defined interface in some runtime. How exactly should it check if that kind of caller is authorized?

Reply:
Do you see anywhere the discussion on user-defined interfaces and if the core shall control the caller there? Me neither — [we discuss the built-in configuration interface that must be invoked by the framework only](https://github.com/exonum/exonum/pull/1458#discussion_r325171020), and the little sense it makes to place a burden of controlling the caller in this case on client code, be it the runtime or the service implementation.

Reply:
Guys. Let's discuss this inside ?